src/persist-client/src/operators/shard_source.rs

```diff
@@ -66,11 +66,10 @@ use crate::{Diagnostics, PersistLocation, ShardId};
 /// using [`timely::dataflow::operators::generic::operator::empty`].
 ///
 /// [advanced by]: differential_dataflow::lattice::Lattice::advance_by
-pub fn shard_source<'g, K, V, T, D, F, DT, G>(
+pub fn shard_source<'g, K, V, T, D, F, DT, G, C>(
     scope: &mut Child<'g, G, T>,
     name: &str,
-    clients: Arc<PersistClientCache>,
-    location: PersistLocation,
+    client: impl Fn() -> C,
```

Comment:
can this just a be a `: Future` itself?

Reply:
I wish! It needs two of them, so the option I saw where: the Future has to be Clone (how?), pass two of of these client futures, or do what I did and make it a `Fn` that generates `Future<Output=PersistClient>`. I think this is dumb though, so definitely open to alternatives!

Reply:
ah i forgot we need to call this more than once, this is the right technique for that!